src/palette.c

```diff
@@ -0,0 +1,987 @@
+static void sub_80718B8(u8 taskId);
+
+ALIGNED(4) EWRAM_DATA u16 gPlttBufferUnfaded[PLTT_BUFFER_SIZE] = {0};
+ALIGNED(4) EWRAM_DATA u16 gPlttBufferFaded[PLTT_BUFFER_SIZE] = {0};
+EWRAM_DATA struct PaletteStruct sPaletteStructs[0x10] = {0};
+EWRAM_DATA struct PaletteFadeControl gPaletteFade = {0};
+static EWRAM_DATA u32 sPlttBufferTransferPending = 0;
+EWRAM_DATA u8 gPaletteDecompressionBuffer[PLTT_DECOMP_BUFFER_SIZE] = {0};
+
+static const struct PaletteStructTemplate gDummyPaletteStructTemplate =
+{
+    .uid = 0xFFFF,
+    .pst_field_B_5 = 1
+};
+
+static const u8 sRoundedDownGrayscaleMap[] =
+{
+     0,  0,  0,  0,  0,
+     5,  5,  5,  5,  5,
+    11, 11, 11, 11, 11,
+    16, 16, 16, 16, 16,
+    21, 21, 21, 21, 21,
+    27, 27, 27, 27, 27,
+    31, 31
+};
+
+void LoadCompressedPalette(const u32 *src, u16 offset, u16 size)
+{
+    LZDecompressWram(src, gPaletteDecompressionBuffer);
+    CpuCopy16(gPaletteDecompressionBuffer, gPlttBufferUnfaded + offset, size);
+    CpuCopy16(gPaletteDecompressionBuffer, gPlttBufferFaded + offset, size);
+}
+
+void LoadPalette(const void *src, u16 offset, u16 size)
+{
+    CpuCopy16(src, gPlttBufferUnfaded + offset, size);
+    CpuCopy16(src, gPlttBufferFaded + offset, size);
+}
+
+void FillPalette(u16 value, u16 offset, u16 size)
+{
+    CpuFill16(value, gPlttBufferUnfaded + offset, size);
+    CpuFill16(value, gPlttBufferFaded + offset, size);
+}
+
+void TransferPlttBuffer(void)
+{
+    if (!gPaletteFade.bufferTransferDisabled)
+    {
+        void *src = gPlttBufferFaded;
+        void *dest = (void *)PLTT;
+        DmaCopy16(3, src, dest, PLTT_SIZE);
+        sPlttBufferTransferPending = 0;
+        if (gPaletteFade.mode == HARDWARE_FADE && gPaletteFade.active)
+            UpdateBlendRegisters();
+    }
+}
+
+u8 UpdatePaletteFade(void)
+{
+    u8 result;
+    u8 dummy = 0;
+
+    if (sPlttBufferTransferPending)
+        return PALETTE_FADE_STATUS_LOADING;
+    if (gPaletteFade.mode == NORMAL_FADE)
+        result = UpdateNormalPaletteFade();
+    else if (gPaletteFade.mode == FAST_FADE)
+        result = UpdateFastPaletteFade();
+    else
+        result = UpdateHardwarePaletteFade();
+    sPlttBufferTransferPending = gPaletteFade.multipurpose1 | dummy;
+    return result;
+}
+
+void ResetPaletteFade(void)
+{
+    u8 i;
+
+    for (i = 0; i < 16; ++i)
+        ResetPaletteStruct(i);
+    ResetPaletteFadeControl();
+}
+
+void ReadPlttIntoBuffers(void)
+{
+    u16 i;
+    u16 *pltt = (u16 *)PLTT;
+
+    for (i = 0; i < PLTT_SIZE / 2; ++i)
+    {
+        gPlttBufferUnfaded[i] = pltt[i];
+        gPlttBufferFaded[i] = pltt[i];
+    }
+}
+
+bool8 BeginNormalPaletteFade(u32 selectedPalettes, s8 delay, u8 startY, u8 targetY, u16 blendColor)
+{
+    u8 temp;
+    u16 color = blendColor;
+
+    if (gPaletteFade.active)
+    {
+        return FALSE;
+    }
+    else
+    {
+        gPaletteFade.deltaY = 2;
+        if (delay < 0)
+        {
+            gPaletteFade.deltaY += (delay * -1);
```

Comment:
you really have to question what gf was smoking here (i expect there was a similar comment in ruby and emerald)